File: java_refactoring/src/main/java/org/example/solution/farmcreator/FishFarmCreator.java
package org.example.solution.farmcreator;

import org.example.solution.FarmType;
import org.example.solution.farmcreator.FarmCreator;

public class FishFarmCreator extends FarmCreator {


    public FishFarmCreator(String farmName) {
        super(farmName, FarmType.fromString("fish"));
    }

    @Override
    public int getValueToAdd() {
        return 1000;
    }

    //farmName should be passed in a constructor
    @Override
    public void createLandType() {
        System.out.println("Creating land for " + this.farmName);
        for (int count = 0; count < FarmCreator.ADD_LAND_COUNT; count++) {
            this.farm.append( "crops");
            System.out.println("adding some crops " + this.farmName); // adding farmName to the log would make the log more informative
            this.numCreated += 2;
        }
    }

    @Override
    public void addPlants() {
        System.out.println("Planting clover " + this.farmName);
    }
}

File: typescript_refactoring/example/solution/FishFarmCreator.ts
import { FarmCreator } from "./FarmCreator";
import { FarmType } from "./FarmType";

export class FishFarmCreator extends FarmCreator {
    public constructor(farmName: string) {
        super(farmName, FarmType.Fish);
    }

    public addPlants(): void {
        console.log(`Planting algae for farm ${this.farmName}`);
    }

    protected createLandType(): void {
        this.addFarmPart("fish");
        console.log(`Adding some fish for farm ${this.farmName}`);
        this.numCreated += 1;
    }

    protected getValueToAdd(): number {
        return 2000;
    }
}
